### PR TITLE
Support per-core runtime args in TTNN generic

### PR DESCRIPTION
**Pull request by vtangTT**
### Ticket
https://github.com/tenstorrent/tt-mlir/issues/4827

### Problem description
D2M doesn't use per-core rt args however external consumers might want to. Need to define `RuntimeArgs` in tablegen and fbs to mirror the definition in `program_descriptors.hpp` in metal

### What's changed
- new `TTNN_CoreRuntimeArgsAttr` that follows the `RuntimeArgs` in [tip of metal](https://github.com/tenstorrent/tt-metal/blob/1c6a7e76d46b2554cc16db6db28fb72a46c132e2/tt_metal/api/tt-metalium/program_descriptors.hpp#L106) that uses `std::pair<CoreCoord, CoreRuntimeArgs>` instead of a 2D vector
- new `CoreRuntimeArgs` in fbs that also mirrors definition in metal
- verifier that checks there aren't rt args defined for overlapping cores
- ttnn->flatbuffer translation support for rt args
- updated generic op runtime converts from the `std::pair<CoreCoord, CoreRuntimeArgs>` model to the [old 2D vector model](https://github.com/tenstorrent/tt-metal/blob/2f70c56f7a5a6561d9a6968a973d356b664cbbf8/tt_metal/api/tt-metalium/program_descriptors.hpp#L105)
    - this will need to get changed after uplift goes into main
- updated `generic_op.mlir` test to pass output tensor address through rt arg (originally passed as a crt arg) to qualify flatbuffer and runtime changes

Before merge:
- [x] wait for [uplift](https://github.com/tenstorrent/tt-mlir/pull/6470)
### Checklist
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: runtime/lib/ttnn/operations/generic/generic_op.cpp [hunk 1]
```diff
@@ -178,6 +178,49 @@ static std::vector<uint32_t> createKernelArgs(
   return coreArgs;
 }
 
+// TODO(vtangTT): Uncomment once program_descriptors.hpp is updated to use
```
Comment (vwellsTT) [inline runtime/lib/ttnn/operations/generic/generic_op.cpp]: will this resolve before merging? If not should probably ticket

File: lib/Dialect/TTNN/IR/TTNNOpsAttrs.cpp [hunk 2]
```diff
@@ -1197,6 +1199,23 @@ ::llvm::LogicalResult verifyCompileTimeArgs(
   return ::llvm::success();
 }
 
+::llvm::LogicalResult verifyRuntimeArgs(
+    ::llvm::function_ref<::mlir::InFlightDiagnostic()> emitError,
+    ::llvm::ArrayRef<mlir::tt::ttnn::CoreRuntimeArgsAttr> rtArgs) {
+  // Check for duplicate CoreCoords
```
Comment (vwellsTT) [inline lib/Dialect/TTNN/IR/TTNNOpsAttrs.cpp]: nit: end with .

File: lib/Dialect/TTNN/IR/TTNNOpsAttrs.cpp [hunk 3]
```diff
@@ -1197,6 +1199,23 @@ ::llvm::LogicalResult verifyCompileTimeArgs(
...
+  llvm::DenseSet<std::pair<uint64_t, uint64_t>> seenCoords;
+  for (auto rtArg : rtArgs) {
+    auto coord = rtArg.getCoreCoord();
```
Comment (vwellsTT) [inline lib/Dialect/TTNN/IR/TTNNOpsAttrs.cpp]: nit: bad auto imo

File: lib/Conversion/D2MToTTNN/D2MToTTNN.cpp [hunk 4]
```diff
@@ -144,6 +144,9 @@ class D2MGenericRewriter : public OpConversionPattern<d2m::GenericOp> {
       }
 
       // Create KernelDescriptor.
+      // Empty rt_args since per-core runtime args not used by D2M.
+      ArrayRef<ttnn::CoreRuntimeArgsAttr> emptyRtArgs;
```
Comment (vwellsTT) [inline lib/Conversion/D2MToTTNN/D2MToTTNN.cpp]: debatably would be cleaner to handle this w/ alternate build func for the KernelAttrs s.t. we can omit passing this? Idk

File: runtime/lib/ttnn/operations/generic/generic_op.cpp (hunk 1 above)
Comment (vtangTT) [reply runtime/lib/ttnn/operations/generic/generic_op.cpp]: will resolve before yes, gonna wait uplift before merge.

**Review by jameszianxuTT — APPROVED** [review]

File: lib/Dialect/TTNN/IR/TTNNOpsAttrs.cpp [hunk 5]
```diff
@@ -1197,6 +1199,23 @@ ::llvm::LogicalResult verifyCompileTimeArgs(
   return ::llvm::success();
 }
 
+::llvm::LogicalResult verifyRuntimeArgs(
```
Comment (jameszianxuTT) [inline lib/Dialect/TTNN/IR/TTNNOpsAttrs.cpp:1202]: Nit, but maybe you can add a xfail test with duplicate CoreCoords to make sure this function actually works? 

It's currently simple enough that it's probably not necessary atm

**Review by xanderchin — APPROVED** [review]
Thanks for the speedy PR

File: lib/Dialect/TTNN/IR/TTNNOpsAttrs.cpp (hunk 3 above)
Comment (vtangTT) [reply lib/Dialect/TTNN/IR/TTNNOpsAttrs.cpp]: fixed

File: lib/Conversion/D2MToTTNN/D2MToTTNN.cpp (hunk 4 above)
Comment (vtangTT) [reply lib/Conversion/D2MToTTNN/D2MToTTNN.cpp]: added custom builders

File: lib/Dialect/TTNN/IR/TTNNOpsAttrs.cpp (hunk 2 above)
Comment (vtangTT) [reply lib/Dialect/TTNN/IR/TTNNOpsAttrs.cpp]: fixed

File: lib/Dialect/TTNN/IR/TTNNOpsAttrs.cpp (hunk 5 above)
Comment (vtangTT) [reply lib/Dialect/TTNN/IR/TTNNOpsAttrs.cpp:1202]: added `test/ttmlir/Silicon/TTNN/n150/generic_op/runtime_args_negative.mlir`
